File: rd-framework/rd-framework-jvm/src/test/kotlin/com/jetbrains/rider/framework/test/cases/RdExtTest.kt
package com.jetbrains.rider.framework.test.cases

import com.jetbrains.rider.framework.*
import com.jetbrains.rider.framework.base.*
import com.jetbrains.rider.framework.impl.RdOptionalProperty
import com.jetbrains.rider.framework.impl.RdProperty
import com.jetbrains.rider.util.lifetime.Lifetime
import com.jetbrains.rider.util.reactive.IProperty
import com.jetbrains.rider.util.reactive.valueOrThrow
import org.testng.annotations.Test
import kotlin.reflect.KClass
import kotlin.test.assertEquals

class RdExtTest : RdTestBase() {
    @Test(enabled = false) // TODO: RIDER-14180
    fun testExtension() {
        val propertyId = 1
        val clientProperty = RdOptionalProperty<DynamicEntity>().static(propertyId)
        val serverProperty = RdOptionalProperty<DynamicEntity>().static(propertyId).slave()

        DynamicEntity.create(clientProtocol)
        DynamicEntity.create(serverProtocol)
        //bound
        clientProtocol.bindStatic(clientProperty, "top")
        serverProtocol.bindStatic(serverProperty, "top")


        clientWire.autoFlush = false
        serverWire.autoFlush = false

        var clientEntity = DynamicEntity(true)
        clientProperty.set(clientEntity)
        clientWire.processAllMessages()

        val serverEntity = DynamicEntity(true)
        serverProperty.set(serverEntity)
        serverWire.processAllMessages()

        //it's new!
        clientEntity = clientProperty.valueOrThrow

        clientEntity.getOrCreateExtension("ext", DynamicExt::class) { DynamicExt("Ext!", "client") }
        clientWire.processAllMessages()
        //client send READY

        val serverExt = serverEntity.getOrCreateExtension("ext", DynamicExt::class) { DynamicExt("", "server") }
        serverWire.processAllMessages()
        //server send READY


        clientWire.processAllMessages()
        //client send COUNTERPART_ACK

        serverWire.processAllMessages()
        //server send COUNTERPART_ACK

        assertEquals("Ext!", serverExt.bar.value)
    }

    private class DynamicEntity(val _foo: RdProperty<Boolean?>) : RdBindableBase() {
        val foo : IProperty<Boolean?> = _foo

        companion object : IMarshaller<DynamicEntity> {
            override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): DynamicEntity {
                return DynamicEntity(RdProperty.read(ctx, buffer, FrameworkMarshallers.Bool.nullable()))
            }

            override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: DynamicEntity) {
                RdProperty.write(ctx, buffer, value._foo)
            }

            override val _type: KClass<*>
                get() = DynamicEntity::class

            fun create(protocol: IProtocol) {
                protocol.serializers.register(DynamicEntity)
            }
        }

        override fun init(lifetime: Lifetime) {
            _foo.bind(lifetime, this, "foo")
        }

        override fun identify(identities: IIdentities, ids: RdId) {
            _foo.identify(identities, ids.mix("foo"))
        }

        constructor(_foo : Boolean?) : this(RdProperty(_foo, FrameworkMarshallers.Bool.nullable()))
    }

    private class DynamicExt(val _bar: RdProperty<String>, private val debugName: String) : RdExtBase(), ISerializersOwner {
        override fun registerSerializersCore(serializers: ISerializers) {}

        override val serializersOwner: ISerializersOwner get() = this

        val bar: IProperty<String> = _bar

        init {
            bindableChildren.add("bar" to _bar)
            _bar.slave()
        }

        companion object : IMarshaller<DynamicExt> {
            override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): DynamicExt {
                throw IllegalStateException()
            }

            override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: DynamicExt) {
                RdProperty.write(ctx, buffer, value._bar)
            }

            override val _type: KClass<*>
                get() = DynamicExt::class

            fun create(protocol: IProtocol) {
                protocol.serializers.register(DynamicExt)
            }
        }

        constructor(_bar: String, debugName: String) : this(RdProperty(_bar, FrameworkMarshallers.String), debugName)
    }

}

File: rd-cpp/src/rd_framework_cpp/src/test/cases/RdExtTest.cpp
#include <gtest/gtest.h>

#include "impl/RdProperty.h"

#include "RdFrameworkTestBase.h"
#include "DynamicEntity.h"
#include "DynamicExt.h"
#include "SocketWireTestBase.h"
#include "ExtProperty.h"

using namespace rd;
using namespace test;
using namespace test::util;

// TEST_F(SocketWireTestBase, testStringExtension) {
//    Protocol serverProtocol = server(socketLifetime);
//    Protocol clientProtocol = client(socketLifetime, serverProtocol);
//
//    RdProperty<int32_t> sp{0}, cp{0};
//    cp.slave();
//
//    init(serverProtocol, clientProtocol, &sp, &cp);
//
//    sp.getOrCreateExtension<std::string>("data", []() { return "Immutable"; });
//
//    cp.getOrCreateExtension<std::string>("data", []() { return "Immutable"; });
//
//    sp.getOrCreateExtension<int>("data", []() { return int(1); }) = 2;
//
//    EXPECT_EQ(cp.get(), 0);
//    EXPECT_EQ(sp.get(), 0);
//
//    sp.set(1);
//    clientScheduler.pump_one_message();
//
//    std::string const &clientExt = cp.getOrCreateExtension<std::string>("data", []() { return "Mutable"; });
//    std::string const &serverExt = sp.getOrCreateExtension<std::string>("data", []() { return "Mutable"; });
//
//    checkSchedulersAreEmpty();
//
//    EXPECT_EQ(clientExt, "Immutable");
//    EXPECT_EQ(serverExt, "Immutable");
//
//    terminate();
//}

/*TEST_F(SocketWireTestBase, DISABLED_testExtension) {
	int property_id = 1;
	int entity_id = 2;
	int32_t foo_id = 3;

	Protocol serverProtocol = server(socketLifetime);
	Protocol clientProtocol = client(socketLifetime, serverProtocol);

	RdProperty<DynamicEntity> clientProperty{DynamicEntity(0)};
	statics(clientProperty, property_id);
	statics(clientProperty.get(), entity_id);
	statics(clientProperty.get().foo, foo_id);

	RdProperty<DynamicEntity> serverProperty{DynamicEntity(0)};
	statics(serverProperty, property_id);
	statics(serverProperty.get(), entity_id);
	statics(serverProperty.get().foo, foo_id);

	serverProperty.slave();

	DynamicEntity::create(&clientProtocol);
	DynamicEntity::create(&serverProtocol);
	//bound
	clientProperty.bind(lifetime, &clientProtocol, static_name);
	serverProperty.bind(lifetime, &serverProtocol, static_name);

	DynamicEntity clientEntity(1);
	clientProperty.set(std::move(clientEntity));
	serverScheduler.pump_one_message();

	DynamicEntity serverEntity(1);
	serverProperty.set(std::move(serverEntity));
	clientScheduler.pump_one_message();

	//it's new!
	auto const &newServerEntity = serverProperty.get();

	DynamicExt const &serverExt = newServerEntity.getOrCreateExtension<DynamicExt>("ext", L"Ext!", L"client");

	clientScheduler.pump_one_message();
	//server send READY

	auto const &newClientEntity = clientProperty.get();
	DynamicExt const &clientExt = newClientEntity.getOrCreateExtension<DynamicExt>("ext", L"", L"server");

	serverScheduler.pump_one_message();
	//client send READY

	clientScheduler.pump_one_message();
	//server send COUNTERPART_ACK

	checkSchedulersAreEmpty();

	EXPECT_EQ(L"Ext!", serverExt.bar.get());
	EXPECT_EQ(L"Ext!", clientExt.bar.get());

	terminate();
}*/

TEST_F(SocketWireTestBase, /*DISABLED_*/ testSlowpokeExtension)
{
	//	int64_t const serialization_hash = 1ll << 40u;

	Protocol serverProtocol = server(socketLifetime);
	Protocol clientProtocol = client(socketLifetime, serverProtocol);

	RdProperty<int> serverProperty{0}, clientProperty{0};
	init(serverProtocol, clientProtocol, &serverProperty, &clientProperty);

	auto const& serverExt = serverProperty.getOrCreateExtension<ExtProperty<std::wstring>>("data", L"SERVER");
	//	serverExt.serializationHash = serialization_hash;

	serverExt.property.set(L"UPDATE");
	serverExt.property.set(L"UPGRADE");

	auto const& clientExt = clientProperty.getOrCreateExtension<ExtProperty<std::wstring>>("data", L"CLIENT");
	//	clientExt.serializationHash = serialization_hash;

	EXPECT_EQ(clientExt.property.get(), L"CLIENT");

	//	clientScheduler.pump_one_message(); //send Ready
	//	serverScheduler.pump_one_message(); //send Ready
	//	serverScheduler.pump_one_message(); //send ReceivedCounterpart
	//	clientScheduler.pump_one_message(); //send ReceivedCounterpart
	// no need in pumping due to synchronous scheduler
	clientScheduler.pump_one_message();	   // send "UPDATE"

	EXPECT_EQ(serverExt.property.get(), L"UPGRADE");
	EXPECT_EQ(clientExt.property.get(), L"UPDATE");

	clientScheduler.pump_one_message();	   // send "UPGRADE"
	checkSchedulersAreEmpty();

	EXPECT_EQ(serverExt.property.get(), L"UPGRADE");
	EXPECT_EQ(clientExt.property.get(), L"UPGRADE");

	terminate();
}
